File: Proj2/golang/pegboard.go
package main

import (
  "fmt"
  "math"
  // "strconv"
)

type Move int
const (
  TOPLEFT Move = iota
  TOPRIGHT
  LEFT
  RIGHT
  BOTTOMLEFT
  BOTTOMRIGHT
)

type MoveDelta struct {
  dRow int
  dCol int
}
var moveDict = map[Move]MoveDelta {
  TOPLEFT     : MoveDelta{-2, -2},
  TOPRIGHT    : MoveDelta{-2, 0},
  LEFT        : MoveDelta{0, -2},
  RIGHT       : MoveDelta{0, 2},
  BOTTOMLEFT  : MoveDelta{2, 0},
  BOTTOMRIGHT : MoveDelta{2, 2},
}


type PegBoard struct {
  boardArray [][]bool
}

func initializeBoard(rows int, row int, col int) *PegBoard {
  pegBoard := new(PegBoard)
  pegBoard.boardArray = make([][]bool, rows)

  for i := 0; i < rows; i++ {
    pegBoard.boardArray[i] = make([]bool, i + 1)
    for j := 0; j <= i; j++ {
      pegBoard.boardArray[i][j] = true
    }
  }

  pegBoard.boardArray[row][col] = false

  return pegBoard
}

/*
 *  
 */
func (pegBoard *PegBoard) setBoard(binary string) {
  rows := int(math.Sqrt(float64(2 * len(binary))))

  board := make([][]bool, rows)

  currRow := 0
  stringPos := 0

  for stringPos < len(binary) {
    board[currRow] = make([]bool, currRow + 1)
    for currCol := 0; currCol <= currRow; currCol++ {
      board[currRow][currCol] = string(binary[stringPos]) == "1"
      stringPos++
    }
    currRow++
  }

  pegBoard.boardArray = board
}


func (pegBoard *PegBoard) canMakeMove(row int, col int, move Move) bool {
  board := pegBoard.boardArray

  // Checks if position is a peg
  if !board[row][col] {
    return false
  }

  destRow := row + moveDict[move].dRow
  destCol := col + moveDict[move].dCol

  // Checks if destination is a valid position on the board
  if !pegBoard.posOnBoard(destRow, destCol) {
    return false
  }

  midRow := (row + destRow) / 2
  midCol := (col + destCol) / 2

  return !board[destRow][destCol] && board[midRow][midCol]
}

func (pegBoard *PegBoard) makeMove(row int, col int, move Move) {
  board := pegBoard.boardArray

  destRow := row + moveDict[move].dRow
  destCol := col + moveDict[move].dCol

  midRow := (row + destRow) / 2
  midCol := (col + destCol) / 2

  board[row][col] = false;
  board[midRow][midCol] = false;
  board[destRow][destCol] = true;
}

func getPieceNumber(row int, col int) int {
  return ((row * (row + 1)) / 2) + col;
}

func (pegBoard *PegBoard) posOnBoard(row int, col int) bool {
  boardArray := pegBoard.boardArray
  return (row >= 0 && row < len(boardArray)) && (col >= 0 && col <= row);
}


func (pegBoard *PegBoard) getBinary() string {
  boardArray := pegBoard.boardArray
  binary := ""

  for i := 0; i < len(boardArray); i++ {
    for j := 0; j <= i; j++ {
      if boardArray[i][j] {
        binary += "1"
      } else {
        binary += "0"
      }
    }
  }

  return binary
}

func (pegBoard *PegBoard) printBoard() {
  boardArray := pegBoard.boardArray

  for i := 0; i < len(boardArray); i++ {
    for j := 0; j <= i; j++ {
      if boardArray[i][j] {
        fmt.Print("1 ")
      } else {
        fmt.Print("0 ")
      }
    }
    fmt.Println()
  }
}

File: Proj2/java/src/com/hw/PegBoard.java
package com.hw;

/**
* Representation of a Triangle Peg Board.
*
* @Author Joonho Kim
*/
public class PegBoard {
  private boolean[][] board;

  public PegBoard(int rows) {
    board = new boolean[rows][];
    for (int i = 0; i < rows; i ++) {
      board[i] = new boolean[i + 1];
    }
  }
  
  public PegBoard(String binary) {
    char[] charArray = binary.toCharArray();
    int rows = (int) Math.sqrt(binary.length() * 2);

    board = new boolean[rows][];

    int currRow = 0;
    int charArrayPos = 0;
    while (charArrayPos < charArray.length) {
      board[currRow] = new boolean[currRow + 1];
      for (int currCol = 0; currCol <= currRow; currCol++) {
        board[currRow][currCol] = (charArray[charArrayPos] == '1');
        charArrayPos++;
      }
      currRow++;
    }
  }

  public void setBoard(String binary) {
    char[] charArray = binary.toCharArray();
    int rows = (int) Math.sqrt(binary.length() * 2);

    board = new boolean[rows][];

    int currRow = 0;
    int charArrayPos = 0;
    while (charArrayPos < charArray.length) {
      board[currRow] = new boolean[currRow + 1];
      for (int currCol = 0; currCol <= currRow; currCol++) {
        board[currRow][currCol] = (charArray[charArrayPos] == '1');
        charArrayPos++;
      }
      currRow++;
    }
  }


  /**
  * Sets up the board with an empty spot at row,col
  */
  public void initializeBoard(int row, int col){
    int rows = getRows();
    if (!posOnBoard(row, col)) {
      throw new IndexOutOfBoundsException("Row/Col is not valid position on the board");
    }
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < i + 1; j++) {
        board[i][j] = true;
      }
    }
    board[row][col] = false;
  }
  
  public static int getPieceNumber(int row, int col) {
    return ((row * (row + 1)) / 2) + col;
  }

  /**
  * Makes a move on piece row,col
  * Assumes that row,col can make the move.
  */
  public void makeMove(int row, int col, PegBoardMove move) {   
    int destRow = row + move.getRowMove();
    int destCol = col + move.getColMove();
    
    int midRow = (destRow + row) / 2;
    int midCol = (destCol + col) / 2;
    
    board[row][col] = false;
    board[midRow][midCol] = false;
    board[destRow][destCol] = true;
  }
  
  /**
  * Checks to see whether a peg at row,col can make the move
  */
  boolean canMakeMove(int row, int col, PegBoardMove move) {
    // piece at row,col is not a peg
    if (!board[row][col]) {
      return false;
    }

    int destRow = row + move.getRowMove();
    int destCol = col + move.getColMove();
    
    // Valid position on Board
    if (!posOnBoard(destRow, destCol)) {
      return false;
    }
    
    // Position in between dest and curr has a piece
    int midRow = (destRow + row) / 2;
    int midCol = (destCol + col) / 2;
    
    // Dest is empty and mid is a peg
    return !board[destRow][destCol] && board[midRow][midCol];
  }
  
  private boolean posOnBoard(int row, int col) {
    return (row >= 0 && row < board.length) && (col >= 0 && col <= row);
  }
    
  public boolean[][] getBoard() {
    return board;
  }
    
  public int getRows() {
    return board.length;
  }
  
  /**
  * Binary Representation of the pegBoard
  * From top to bottom, left to right, we create a string where
  * a piece is 1 and a empty space is 0.
  * This is used as a key to our HashMap of PegBoard to StateNextMove
  */
  public String getBinary() {
    String bin = "";
    for (boolean[] bArr : board) {
      for (boolean b : bArr) {
        bin += (b) ? "1" : "0";
      }
    }
    return bin;
  }
  
  public boolean equals(Object o) {
    if (o instanceof PegBoard) {
      return getBinary().equals(((PegBoard) o).getBinary());
    }
    return false;
  }
  
  public String toString() {
    String output = "";
    for (boolean[] b : board) {
      for (boolean bool : b) {
        output += (bool) ? "1" : "0";
      }
      output += "\n";
    }
    return output;
  }
  
  public void printBoard() {
    System.out.println(toString());
  }
}